codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/SmithyEndpointGenerator.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.go.codegen.integration;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Consumer;
+import software.amazon.smithy.codegen.core.SymbolProvider;
+import software.amazon.smithy.go.codegen.EndpointGenerator;
+import software.amazon.smithy.go.codegen.GoSettings;
+import software.amazon.smithy.go.codegen.GoWriter;
+import software.amazon.smithy.go.codegen.SymbolUtils;
+import software.amazon.smithy.go.codegen.TriConsumer;
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.rulesengine.language.EndpointRuleSet;
+import software.amazon.smithy.rulesengine.traits.EndpointRuleSetTrait;
+import software.amazon.smithy.utils.ListUtils;
+import software.amazon.smithy.utils.SetUtils;
+
+public class SmithyEndpointGenerator implements GoIntegration {
+
+    public static final String PUBLIC_PARAMETERS_NAME = "EndpointParameters";
+
+    @Override
+    public void writeAdditionalFiles(
+            GoSettings settings,
+            Model model,
+            SymbolProvider symbolProvider,
+            TriConsumer<String, String, Consumer<GoWriter>> writerFactory) {
+
+        var serviceShape = settings.getService(model);
+
+        var publicOptionsType = SymbolUtils.createValueSymbolBuilder(PUBLIC_PARAMETERS_NAME).build();
+
+        var generator = EndpointGenerator.builder()
+                .publicParametersType(publicOptionsType)
+                .build();
+
+
+        Optional<EndpointRuleSet> rulesetOpt = Optional.empty();
+        var ruleSetTraitOpt = serviceShape.getTrait(EndpointRuleSetTrait.class);
+        if (ruleSetTraitOpt.isPresent()) {
+            rulesetOpt = Optional.of(EndpointRuleSet.fromNode(ruleSetTraitOpt.get().getRuleSet()));
+        }
+
+        Optional<EndpointRuleSet> finalRulesetOpt = rulesetOpt;
+        writerFactory.accept("endpoints.go", settings.getModuleName(), (w) -> {
+                    if (finalRulesetOpt.isPresent()) {
+                        w.write("$W", generator.generateParameters(finalRulesetOpt.get()));
```

Comment:
**suggestion**: This branch between `generateParameters` and `generateEmptyParameters` feels like it should just be internalized in the parameters generator or `EndpointGenerator`.

Reply:
yah i dont think its perfect to have the decision point between empty or not empty parameters be in the class up top. the original reason for leaving it here was so that the the `EndpointParameters` generator could be ignorant of a ruleset, which i think its more appropriate for parameter extraction out of a ruleset to be in the top-level `SmithyEndpointGenerator`. But i get how when you immediately see logic thats really only related to parameter generation, it feels leaky. so ill make the change.